Claimable balance example contract (demonstrates timelock).

dmkozh opened the pull request:
What
Example ledger timestamp access (https://github.com/stellar/soroban-examples/issues/72)

As the relevant part is pretty small/trivial, I also took this as an opportunity to try a slightly different storage style and also tried to add more test coverage than the other examples.

Why
Adding more Soroban examples

Known limitations
N/A

sisuresh commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+) {
+    let client = TokenClient::new(&e, token_id);
+    let nonce = client.nonce(&get_contract_id(&e));
+    client.xfer_from(&Signature::Contract, &nonce, &from, &to, &amount);
```
The nonce for the Contract Signature type is expected to be 0, so you can just pass that in instead of calling the nonce function.

sisuresh replied:
I just realized the `single_offer_xfer_from` does the same thing. I'll make sure to fix that.

dmkozh replied:
Both calls use contract auth, so only the first one can be 0. I guess we could use 0 for this function, but I'm not sure it's reasonable for a helper function that doesn't necessarily know when is it going to be called.

sisuresh replied:
The auth sdk expects the passed in nonce to be [zero](https://github.com/stellar/rs-soroban-sdk/blob/6d129bf6a9250d54d67ee851ea027d4a4f892673/soroban-auth/src/lib.rs#L128) for `Signature::Contract` because it isn't used, so there's no reason to do this cross contract call. 

The way the nonce is passed into `check_auth` isn't ideal because of this. Maybe our solution to https://github.com/stellar/rs-soroban-sdk/issues/533 will improve this as well.

dmkozh replied:
I see, that's a bit confusing indeed. Changed nonce to zero.

sisuresh commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+fn transfer_to(e: &Env, token_id: &BytesN<32>, to: &Identifier, amount: &BigInt) {
+    let client = TokenClient::new(&e, token_id);
+    let nonce = client.nonce(&get_contract_id(&e));
+    client.xfer(&Signature::Contract, &nonce, to, amount);
```
Same as comment above. You can pass in 0.

dmkozh replied:
We use contract nonce, hence this isn't 0, is it? Strictly speaking it should be 1 for this call when things go as intended.

dmkozh commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+        check_auth(
+            &env,
+            &NonceForSignature(from),
+            BigInt::zero(&env),
```
That's intentional - the contract only supports a single deposit and a single claim. Note, that nonce is per-account, it's not per-contract (this is also explained in the comment).

sisuresh replied:
Wow I'm not sure how I missed that comment...

dmkozh commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,212 @@
+        check_auth(
+            &env,
+            &NonceForSignature(claimant),
+            BigInt::zero(&env),
```
The comment above says that depositor can't be claimant for simplicity. Otherwise this should work fine, shouldn't it? Otherwise, we'd need to expose nonce and pass it explicitly to the calls, which seems unnecessary for a one-shot contract.

leighmcculloch commented on timelock/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+doctest = false
+
+[features]
+default = ["export"]
+export = []
```
❗ I don't think these features are needed for this example. We've been keeping them out of examples that don't need them. Can you remove them?
```suggestion
```

dmkozh replied:
Done.

leighmcculloch commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,210 @@
+// This contract demonstrates 'timelock' concept and implements a
+// greatly simplified Claimable Balance (similar to
+// https://developers.stellar.org/docs/glossary/claimable-balance).
+// The contract allows to deposit some amount of token and allow another
+// account(s) claim it before or after provided time point.
```
:bulb: The `//!` prefix on a comment acts as a comment for the module / crate, and shows up in docs. We don't generate docs for these, but feels appropriate to use that style of comment for the comment I see here, which seems to describe the crate as a whole.
```suggestion
//! This contract demonstrates 'timelock' concept and implements a
//! greatly simplified Claimable Balance (similar to
//! https://developers.stellar.org/docs/glossary/claimable-balance).
//! The contract allows to deposit some amount of token and allow another
//! account(s) claim it before or after provided time point.
```

dmkozh replied:
Done.

leighmcculloch commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,210 @@
+        token: BytesN<32>,
+        amount: BigInt,
+        claimants: Vec<Identifier>,
+        time_bound_kind: TimeBoundKind,
+        timestamp: u64,
```
💡 These two fields seem to be the same thing as the `TimeBound` type, which suggests you could use that type here.
```suggestion
        time_bound: TimeBound,
```

dmkozh replied:
Done.

leighmcculloch commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,210 @@
+        let claimant_id = claimant.get_identifier(&env);
+        let claimants = &claimable_balance.claimants;
+        // Check if claimaint is one of the allowed claimants.
+        // `contains` is not implemented for Soroban vec (yet).
```
❗ `Vec::contains` is implemented now, so you should be able to use it. Let me know if you have any issues with it.

dmkozh replied:
Done.

leighmcculloch commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,210 @@
+    check_auth, NonceAuth, {Identifier, Signature},
+};
+use soroban_sdk::{contractimpl, contracttype, BigInt, BytesN, Env, IntoVal, Symbol, Vec};
+use soroban_token_contract as token;
+use token::TokenClient;
```
💡 We no longer need to have the token contract as a dependency outside of tests if you use the generated client. If you add the following you can then reference the client as `token::ContractClient`. This helps to illustrate how others might interact with contracts that aren't the token contract that don't have the hand coded machinery.

```suggestion
mod token {
    soroban_sdk::contractimport!(file = "../soroban_token_contract.wasm");
}
```

dmkozh replied:
Done, though I don't fully understand this. Will we somehow deduce that this is token contract and switch to the native version?

leighmcculloch replied:
You're already importing the client for the non-native contract and using it. For how this will work with the native contract, I think we'll probably package the specification into a file that can be imported here just like a wasm file, but it won't contain the program.'

dmkozh replied:
I see, thanks for the explanation.

leighmcculloch reviewed (commented):
Couple small asks (❗). Can you make those tweaks then we'll merge?

Also left a couple suggestions (💡) if you agree.

leighmcculloch commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    pub fn claim(env: Env, claimant: Signature) {
+        if !is_initialized(&env) {
+            panic!("contract hasn't been initialized");
+        }
```
This check looks unnecessary. The following line will fail if there is no balance deposited.

dmkozh replied:
Removed the check.

leighmcculloch commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+}
+
+// Contract usage pattern (pseudocode):
+// 1. Depositor calls `token.approve(depositor_auth, claimable_balance_contract, 100)`
+//    to allow contract to withdraw the needed amount of token.
+// 2. Depositor calls `deposit(depositor_auth, token_id, 100, claimants, time bound)`. Contract
+//    withdraws the provided token amount and stores it until one of the claimants
+//    claims it.
```
It's unfortunate steps 1 and 2 have to be separate steps, but afaict it has to be this way without introducing significant complexity into the token contract and how it verifies signatures. cc @tamirms curious if you think there's other ways to do this that don't require two steps.

tamirms replied:
@leighmcculloch The only way I can think of reducing the number of transactions is through something like [EIP 2612](https://eips.ethereum.org/EIPS/eip-2612) which introduces an extension to the erc20 standard that allows you to do the approval step off-chain through a signing process. There are [several tokens](https://help.1inch.io/en/articles/5435386-permit-712-signed-token-approvals-and-how-they-work-on-1inch) which have implemented  [EIP 2612](https://eips.ethereum.org/EIPS/eip-2612) including DAI and UNI.

dmkozh replied:
IIUC supporting call stack in auth would help resolving this - I feel like it's a useful feature to have in general.

leighmcculloch replied:
I don't think we need to do either of those options right now for this example, the example is good as is. But maybe this example is a good test bed for complex auth scenarios using the soroban-auth crate, once we resolves some of the issues we have open for it.

leighmcculloch reviewed (approved):
Looks good to me. I left some comments, but nothing that needs addressing before merging, other than maybe removing the duplicate check if you agree. Let me know when this should be merged.

dmkozh commented:
> Looks good to me. I left some comments, but nothing that needs addressing before merging, other than maybe removing the duplicate check if you agree. Let me know when this should be merged.

I've removed the duplicate check and squashed fixups - this should be ready for merging now.

tamirms commented on timelock/src/lib.rs:
```diff
@@ -0,0 +1,196 @@
+        let from_id = from.get_identifier(&env);
+        // Authenticate depositor with nonce of zero, so that this may
+        // be successfully called just once.
+        check_auth(
```
what happens if deposit is called twice by different depositors? it seems that the second call will overwrite `DataKey::Balance`?

leighmcculloch replied:
There's a check at the top of `deposit`, a call to `is_initialized`, which could probably have a better name and be named something like `has_deposit`.

tamirms replied:
@leighmcculloch I see `is_initialized()` is called from the `claim()` but I don't think it's called from `deposit()`

leighmcculloch replied:
Ah, wait, @dmkozh I think the is_initialized got removed from teh wrong function. It should have been removed from claim, not from deposit.

tamirms replied:
btw, I think it's possible to have an alternative design where multiple claimable balances are stored within 1 contract rather than having a separate contract per claimable balance. 

This could be implemented by adding an extra parameter to the `deposit()` and `claim()` functions which represents the claimable balance's id (or maybe the id can be derived by hashing the claimable balance + a salt). The balance would be mapped to the id in storage.  I think the deposit flow would be cheaper in this design because the depositor would not need to deploy a new smart contract.

dmkozh replied:
> I think it's possible to have an alternative design where multiple claimable balances are stored within 1 contract rather than having a separate contract per claimable balance.

Indeed, I've considered that option as well. However, it goes beyond the point of the example (which is about interacting with the ledger state in the first place) and is more complex. FWIW, the same principle applies to most of the examples (e.g. one could have a contract that manages several liquidity pools or offers). This has been discussed on discord as well:  https://discord.com/channels/897514728459468821/1013959254245388430

dmkozh replied:
One important con of a 'single contract-multiple entities' approach is that we need to either block on a single autoincrement entry for producing the ids or need to introduce some good way of producing the unique ids deterministically (hash is not enough as one may have multiple entries with the same content).
